### PR TITLE
Feature 74 weather regime plots

### LINKED ISSUE #74: Create Plots for S2S Weather Regime Analysis
## Describe the Task ##
Add the 3 plots used in the Weather Regime analysis.  This includes a line plot in elbow.py, the plots in Calc_EOF.py which need to be converted from basemap to cartopy, and the map in K_means.py

### Time Estimate ###
Generally about 3 days of work

### Sub-Issues ###
Consider breaking the task down into sub-issues.
- [x] Create the elbow plot
- [x] Create the EOF plot
- [x] Create the K means plot

### Relevant Deadlines ###
As soon as possible

### Funding Source ###
UIUC 7790971

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [x] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Task Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_74_weather_regime_plots`
- [x] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by CPKalb**
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
   The testing performed was running the use case to ensure that the plotting worked and came out the same as the original user's version

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
Run the weather regime use case (s2s/UserScript_obsERA_obsOnly_WeatherRegime.conf) which is in branch 569 of METplus.  The output should be 3 plots, obs_eof.png, obs_kmeans.png and obs_elbow.png

- [x] Do these changes include sufficient documentation and testing updates? **[Yes or No]**
 Yes

- [x] Will this PR result in changes to the test suite? **[Yes or No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
No, the weather regime use case changed the test suite

## Pull Request Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, and **Milestone**
- [x] After submitting the PR, select **Linked Issues** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by bikegeek** [comment]
Where is the data located?  I generated the docs for this use case and it only indicates:

- Forecast dataset: GEFS Forecast 500 mb height.

- Observation dataset: ERA Reanlaysis 500 mb height.

I looked under the /d1/projects/METplus/METplus_Data/model_applications/s2s on kiowa and eyewall and didn't find any data that match with the above description.

**Comment by bikegeek** [comment]
Import error when initially running without data:
ModuleNotFoundError: No module named 'eofs'

The error is in this file:
METplus/parm/use_cases/model_applications/s2s/UserScript_obsERA_obsOnly_WeatherRegime/WeatherRegime.py

File: metplotpy/contributed/weather_regime/plot_weather_regime.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import matplotlib.colors as colors
+from pylab import *
+import cmocean
+from cartopy.mpl.gridliner import LONGITUDE_FORMATTER, LATITUDE_FORMATTER
+import cartopy.crs as ccrs
+import cartopy.feature as cfeature
+from cartopy.util import add_cyclic_point
+import numpy as np
+import sys
+sys.path.insert(0, "/glade/u/home/kalb/UIUC/METcalcpy/")
```
Comment (bikegeek) [inline metplotpy/contributed/weather_regime/plot_weather_regime.py]: path is cheyenne-specific, use relative path with the appropriate number of '../' to get to intended directory

File: metplotpy/contributed/weather_regime/plot_weather_regime.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+import matplotlib
+matplotlib.use('agg')
+import matplotlib.pyplot as plt
+import matplotlib.colors as colors
+from pylab import *
+import cmocean
```
Comment (bikegeek) [inline metplotpy/contributed/weather_regime/plot_weather_regime.py:6]: need to add cmocean to METplotpy and METcalcpy documentation for required, installed packages.

File: metplotpy/contributed/weather_regime/plot_weather_regime.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        plt.title('EOF '+str(i+1)+' ('+str(round(variance_fractions[i],2))+' %)')
+
+    plt.tight_layout()
+    fmt='png'
+    full_output_plot = output_plotname + "." + fmt
+    plt.savefig(full_output_plot,format=fmt,dpi=400, bbox_inches='tight')
+
+
+def plot_K_means(inputi,wrnum,lon,lat,perc,output_plotname,plevels):
+
```
Comment (bikegeek) [inline metplotpy/contributed/weather_regime/plot_weather_regime.py:63]: missing python docs with description of input variables.

File: metplotpy/contributed/weather_regime/plot_weather_regime.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    plt.ylabel('Sum_of_squared_distances')
+    plt.title(plot_title)
+
+    fmt = 'png'
+    full_output_plot = output_plotname + "." + fmt
+    plt.savefig(full_output_plot,format=fmt,dpi=400, bbox_inches='tight')
+
+
+def plot_eof(eof,wrnum,variance_fractions,lons,lats,output_plotname,plevels):
+
```
Comment (bikegeek) [inline metplotpy/contributed/weather_regime/plot_weather_regime.py:30]: missing python docs with description of input variables, eg.

'''
Args:
   @param eof:  The xyz abc
   @param wrnum:  some number
 
'''

**Comment by CPKalb** [comment]
The data is on mohawk.   I had to have eofs installed on cheyenne to use the script.  I'm guessing we may need to add it to METplotpy/METcalcpy?

File: metplotpy/contributed/weather_regime/plot_weather_regime.py (hunk 1 above)
Comment (CPKalb) [reply metplotpy/contributed/weather_regime/plot_weather_regime.py]: I'll fix this as soon as cheyenne is back up

**Review by bikegeek — CHANGES_REQUESTED** [review]
There are methods that are missing pydoc comments (suggestions are made in comments).

**Review by bikegeek — CHANGES_REQUESTED** [review]
The template for the issue needs to be filled in with all relevant information:
https://github.com/dtcenter/METplotpy/issues/74

File: metplotpy/contributed/weather_regime/plot_weather_regime.py (hunk 2 above)
Comment (bikegeek) [reply metplotpy/contributed/weather_regime/plot_weather_regime.py:6]: I added cmocean to the Users Guide for METplotpy and METcalcpy in the develop branch

File: metplotpy/contributed/weather_regime/plot_weather_regime.py (hunk 3 above)
Comment (CPKalb) [reply metplotpy/contributed/weather_regime/plot_weather_regime.py:63]: I'm working on the docs right now

File: metplotpy/contributed/weather_regime/plot_weather_regime.py (hunk 1 above)
Comment (CPKalb) [reply metplotpy/contributed/weather_regime/plot_weather_regime.py]: This should now be fixed

File: metplotpy/contributed/weather_regime/plot_weather_regime.py (hunk 4 above)
Comment (CPKalb) [reply metplotpy/contributed/weather_regime/plot_weather_regime.py:30]: The docs are in a README file now in github.  Are there other docs needed?

**Comment by CPKalb** [comment]
I think I have most of the template filled out for issue 74, though there are not unit tests for this code yet.

**Review by bikegeek — CHANGES_REQUESTED** [review]
- Errors encountered when running on 'kiowa', please see log file: 

/d1/personal/minnawin/METplus/output/logs/metplus.log.20210316164537
(permissions updated for you to access)

- running with command:
_run_metplus.py -c /d1/personal/minnawin/metplotpy_PR74/METplus/parm/use_cases/model_applications/s2s/UserScript_obsERA_obsOnly_WeatherRegime.conf -c ./user_system.conf_

- According to the log, it looks like MET command returned a non-zero code, but the use case conf file is NOT invoking any MET tools. 

- installed eofs, metplus and metplotpy in conda environment (used the feature_74_weather_regime_plots version of METplotpy)

**Comment by CPKalb** [comment]
Hi Minna:
Ah, I had a bad path in the use case that George found.  That error was the same one he was getting.  I pushed an update to branch 569 of METplus to fix this error.  I should probably add an error message so it's more clear to the user what is going on in this case.  But, I think you will just need to pull the new version of branch 569 to get it to run.

**Review by bikegeek — APPROVED** [review]
Successfully created three expected plots on 'kiowa', using the conda environment created by George.

**Pull request merged**